File: controller/src/datasets.py
# add the given `dataset` to `envFroms` and return it
def add_dataset(dataset: str, envFroms):
    if envFroms is None:
        envFroms = []

    envFroms.append({
        "secretRef": {
            "name": dataset
        },
        "prefix": dataset + "_"
    })

    return envFroms

def prepare_dataset_labels(application):
    envFroms = []
    volumes = []
    volumeMounts = []

    if "datasets" in application["spec"]:
        idx = 0
        for dataset in application["spec"]["datasets"]:
            name = dataset["name"]
            if "nfs" in dataset:
                volumes.append({
                    "name": name,
                    "nfs": {
                        "path": dataset["nfs"]["path"],
                        "server": dataset["nfs"]["server"],
                    }
                })
                volumeMounts.append({
                    "name": name,
                    "mountPath": dataset["mountPath"] if "mountPath" in dataset else f"/mnt/datasets/{name}",
                })
            elif "pvc" in dataset:
                volumes.append({
                    "name": name,
                    "persistentVolumeClaim": {
                        "claimName": dataset["pvc"]["claimName"],
                    }
                })
                volumeMounts.append({
                    "name": name,
                    "mountPath": dataset["mountPath"] if "mountPath" in dataset else f"/mnt/datasets/{name}",
                })
            elif "s3" in dataset:
                envFroms.append({
                    "secretRef": {
                        "name": dataset["s3"]["secret"]
                    },
                    "prefix": dataset["s3"]["envPrefix"]
                })

    if len(volumes) == 0 and len(volumeMounts) == 0 and len(envFroms) == 0:
        return None, None, None
    else:
        return volumes, volumeMounts, envFroms

File: pkg/fe/linker/datasets.go
package linker

import (
	"lunchpail.io/pkg/fe/linker/helm"
	"lunchpail.io/pkg/fe/linker/yaml/queue"
	"lunchpail.io/pkg/ir/hlir"
)

type nfs struct {
	Path   string `json:"path"`
	Server string `json:"server"`
}

type pvc struct {
	ClaimName string `json:"claimName"`
}

type volume struct {
	Name                  string `json:"name"`
	Nfs                   *nfs `json:"nfs,omitempty"`
	PersistentVolumeClaim *pvc `json:"persistentVolumeClaim,omitempty"`
}

type volumeMount struct {
	Name      string `json:"name"`
	MountPath string `json:"mountPath,omitempty"`
}

type secretRef struct {
	Name string `json:"name"`
}

type envFrom struct {
	SecretRef secretRef `json:"secretRef"`
	Prefix    string `json:"prefix,omitempty"`
}

func envForQueue(queueSpec queue.Spec) envFrom {
	return envFrom{secretRef{queueSpec.Name}, queueSpec.Name + "_"}
}

func datasets(app hlir.Application, queueSpec queue.Spec) ([]volume, []volumeMount, []envFrom, error) {
	volumes := []volume{}
	volumeMounts := []volumeMount{}
	envFroms := []envFrom{envForQueue(queueSpec)}

	for _, dataset := range app.Spec.Datasets {
		name := dataset.Name

		if dataset.Nfs.Server != "" {
			v := volume{}
			v.Name = name
			v.Nfs = &nfs{dataset.Nfs.Server, dataset.Nfs.Path}
			volumes = append(volumes, v)
			volumeMounts = append(volumeMounts, volumeMount{name, dataset.MountPath})
		}
		if dataset.Pvc.ClaimName != "" {
			v := volume{}
			v.Name = name
			v.PersistentVolumeClaim = &pvc{dataset.Pvc.ClaimName}
			volumes = append(volumes, v)
			volumeMounts = append(volumeMounts, volumeMount{name, dataset.MountPath})
		}
		if dataset.S3.Secret != "" {
			envFroms = append(envFroms, envFrom{secretRef{dataset.S3.Secret}, dataset.S3.EnvPrefix})
		}
	}

	return volumes, volumeMounts, envFroms, nil
}

func datasetsB64(app hlir.Application, queueSpec queue.Spec) (string, string, string, error) {
	volumes, volumeMounts, envFroms, err := datasets(app, queueSpec)
	if err != nil {
		return "", "", "", err
	}

	volumesB64, err := helm.ToJsonB64(volumes)
	if err != nil {
		return "", "", "", err
	}

	volumeMountsB64, err := helm.ToJsonB64(volumeMounts)
	if err != nil {
		return "", "", "", err
	}

	envFromsB64, err := helm.ToJsonB64(envFroms)
	if err != nil {
		return "", "", "", err
	}

	return volumesB64, volumeMountsB64, envFromsB64, nil
}
